### PR TITLE
openai[minor]: implement langchain-openai package

### PR DESCRIPTION
**Pull request by efriis**
Todo

- [x] copy over integration tests
- [x] update docs with new instructions in #15513 
- [x] add linear ticket to bump core -> community, community->langchain, and core->openai deps
- [ ] (optional): add `pip install langchain-openai` command to each notebook using it
- [x] Update docstrings to not need `openai` install
- [x] Add serialization
- [x] deprecate old models

Contributor steps:

- [x] Add secret names to manual integrations workflow in .github/workflows/_integration_test.yml
- [x] Add secrets to release workflow (for pre-release testing) in .github/workflows/_release.yml

Maintainer steps (Contributors should not do these):

- [x] set up pypi and test pypi projects
- [x] add credential secrets to Github Actions
- [ ] add package to conda-forge

Functional changes to existing classes:

- now relies on openai client v1 (1.6.1) via concrete dep in langchain-openai package

Codebase organization

- some function calling stuff moved to `langchain_core.utils.function_calling` in order to be used in both community and langchain-openai

### CODE REVIEW

File: libs/community/tests/integration_tests/llms/test_openai.py
Comment (efriis) [inline libs/community/tests/integration_tests/llms/test_openai.py:1]: deleted integration tests that were duplicates of unit tests

File: libs/langchain/langchain/chains/openai_functions/base.py
Comment (efriis) [inline libs/langchain/langchain/chains/openai_functions/base.py:1]: needs re-exports

**Review by hwchase17 — COMMENTED** [review]
_utils.py empty file?

**Review by hwchase17 — COMMENTED** [review]
are there any functional changes from existing classes? if yes, may be helpful to annotate in some way

**Comment by efriis** [comment]
Added to PR description

File: libs/core/langchain_core/load/mapping.py
Comment (efriis) [inline libs/core/langchain_core/load/mapping.py:1]: As discussed this morning

File: libs/community/langchain_community/utils/openai_functions.py
Comment (efriis) [inline libs/community/langchain_community/utils/openai_functions.py:1]: Moved function calling stuff to core/utils/function_calling

File: libs/community/tests/unit_tests/llms/test_openai.py
Comment (efriis) [inline libs/community/tests/unit_tests/llms/test_openai.py:1]: Moved this one from integration tests

File: libs/core/langchain_core/utils/function_calling.py
Comment (efriis) [inline libs/core/langchain_core/utils/function_calling.py:1]: Moved from 

- community/utils/openai_functions
- langchain/chains/openai_functions

and aliased

File: libs/langchain/langchain/chains/openai_functions/base.py
Comment (efriis) [inline libs/langchain/langchain/chains/openai_functions/base.py:1]: moved to core/utils/function_calling

File: libs/partners/openai/README.md
Comment (efriis) [inline libs/partners/openai/README.md:1]: Fill this out

File: libs/partners/openai/langchain_openai/chat_models.py [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+            "tool_call_id": message.tool_call_id,
+        }
+    else:
+        raise TypeError(f"Got unknown type {message}")
+    if "name" in message.additional_kwargs:
+        message_dict["name"] = message.additional_kwargs["name"]
+    return message_dict
+
+
+async def acompletion_with_retry(
```
Comment (efriis) [inline libs/partners/openai/langchain_openai/chat_models.py:149]: remove to simplify

File: libs/partners/openai/langchain_openai/chat_models.py [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+    """
+
+    @property
+    def lc_secrets(self) -> Dict[str, str]:
+        return {"openai_api_key": "OPENAI_API_KEY"}
+
+    @classmethod
+    def get_lc_namespace(cls) -> List[str]:
+        """Get the namespace of the langchain object."""
+        return ["langchain", "chat_models", "openai"]
```
Comment (efriis) [inline libs/partners/openai/langchain_openai/chat_models.py:211]: retain old serialization path

File: libs/partners/openai/langchain_openai/chat_models.py [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+            "stream": self.streaming,
+            "n": self.n,
+            "temperature": self.temperature,
+            **self.model_kwargs,
+        }
+        if self.max_tokens is not None:
+            params["max_tokens"] = self.max_tokens
+        return params
+
+    def completion_with_retry(
```
Comment (efriis) [inline libs/partners/openai/langchain_openai/chat_models.py:374]: remove to simplify

File: libs/partners/openai/langchain_openai/llms.py [hunk 4]
```diff
@@ -0,0 +1,822 @@
...
+async def acompletion_with_retry(
+    llm: BaseOpenAI,
+    run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+    **kwargs: Any,
+) -> Any:
+    """Use tenacity to retry the async completion call."""
+    return await llm.async_client.create(**kwargs)
+
+
+class BaseOpenAI(BaseLLM):
```
Comment (efriis) [inline libs/partners/openai/langchain_openai/llms.py:84]: Removed serialization from this

File: libs/partners/openai/langchain_openai/llms.py [hunk 5]
```diff
@@ -0,0 +1,822 @@
...
+    return llm.client.create(**kwargs)
+
+
+async def acompletion_with_retry(
+    llm: BaseOpenAI,
+    run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+    **kwargs: Any,
+) -> Any:
+    """Use tenacity to retry the async completion call."""
+    return await llm.async_client.create(**kwargs)
```
Comment (efriis) [inline libs/partners/openai/langchain_openai/llms.py:81]: remove to simplify

File: libs/partners/openai/langchain_openai/llms.py [hunk 6]
```diff
@@ -0,0 +1,822 @@
...
+from langchain_core.language_models.llms import BaseLLM
+from langchain_core.outputs import Generation, GenerationChunk, LLMResult
+from langchain_core.pydantic_v1 import Field, root_validator
+from langchain_core.utils import get_from_dict_or_env, get_pydantic_field_names
+from langchain_core.utils.utils import build_extra_kwargs
+
+logger = logging.getLogger(__name__)
+
+
+def update_token_usage(
```
Comment (efriis) [inline libs/partners/openai/langchain_openai/llms.py:39]: make protected

File: libs/partners/openai/langchain_openai/llms.py [hunk 7]
```diff
@@ -0,0 +1,822 @@
...
+        .. code-block:: python
+
+            from langchain_community.llms import OpenAI
+            openai = OpenAI(model_name="gpt-3.5-turbo-instruct")
+    """
+
+    @classmethod
+    def get_lc_namespace(cls) -> List[str]:
+        """Get the namespace of the langchain object."""
+        return ["langchain", "llms", "openai"]
```
Comment (efriis) [inline libs/partners/openai/langchain_openai/llms.py:635]: kept serialization path same

File: libs/partners/openai/langchain_openai/llms.py [hunk 8]
```diff
@@ -0,0 +1,822 @@
...
+
+    @property
+    def _invocation_params(self) -> Dict[str, Any]:
+        return {**{"model": self.model_name}, **super()._invocation_params}
+
+
+class AzureOpenAI(BaseOpenAI):
```
Comment (efriis) [inline libs/partners/openai/langchain_openai/llms.py:642]: Discuss whether we want separate package for azure. Given SDK is openai one, probably makes sense to keep here?

File: libs/partners/openai/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  "I",  # isort
+]
+
+[tool.mypy]
+disallow_untyped_defs = "True"
+
+[[tool.mypy.overrides]]
+module = "transformers"
+ignore_missing_imports = true
+
```
Comment (efriis) [inline libs/partners/openai/pyproject.toml:68]: ignore transformers for custom tokenizer feature in embeddings

File: libs/partners/openai/tests/integration_tests/test_chat_models.py
Comment (efriis) [inline libs/partners/openai/tests/integration_tests/test_chat_models.py:1]: Leaving `scheduled` flags for future use

File: libs/partners/openai/tests/integration_tests/test_embeddings.py
Comment (efriis) [inline libs/partners/openai/tests/integration_tests/test_embeddings.py:1]: Missing

File: libs/partners/openai/tests/unit_tests/test_embeddings.py
Comment (efriis) [inline libs/partners/openai/tests/unit_tests/test_embeddings.py:1]: Missing

File: libs/partners/openai/tests/unit_tests/test_llms.py
Comment (efriis) [inline libs/partners/openai/tests/unit_tests/test_llms.py:1]: Missing

File: libs/community/langchain_community/chat_models/openai.py [hunk 10]
```diff
@@ -143,6 +144,7 @@ def _convert_delta_to_message_chunk(
         return default_class(content=content)
 
 
+@deprecated(since="0.1.0", removal="0.2.0", alternative="langchain_openai.ChatOpenAI")
```
Comment (efriis) [inline libs/community/langchain_community/chat_models/openai.py:147]: causes a decent number of deprecation warnings in unit tests, but still important to test community one before removal.

File: libs/community/langchain_community/llms/openai.py [hunk 11]
```diff
@@ -953,6 +956,7 @@ def lc_attributes(self) -> Dict[str, Any]:
         }
 
 
+@deprecated(since="0.1.0", removal="0.2.0", alternative="langchain_openai.ChatOpenAI")
```
Comment (efriis) [inline libs/community/langchain_community/llms/openai.py:959]: Happy to discuss. This isn't a 1:1 replacement in 0.2

File: libs/partners/openai/langchain_openai/embeddings/base.py [hunk 12]
```diff
@@ -0,0 +1,523 @@
...
+            default=default_api_version,
+        )
+        # Check OPENAI_ORGANIZATION for backwards compatibility.
+        values["openai_organization"] = (
+            values["openai_organization"]
+            or os.getenv("OPENAI_ORG_ID")
+            or os.getenv("OPENAI_ORGANIZATION")
+        )
+        if values["openai_api_type"] in ("azure", "azure_ad", "azuread"):
+            warnings.warn(
```
Comment (efriis) [inline libs/partners/openai/langchain_openai/embeddings/base.py:212]: make raise

File: libs/partners/openai/langchain_openai/embeddings/base.py [hunk 13]
```diff
@@ -0,0 +1,523 @@
...
+            default=default_api_version,
+        )
+        # Check OPENAI_ORGANIZATION for backwards compatibility.
+        values["openai_organization"] = (
+            values["openai_organization"]
+            or os.getenv("OPENAI_ORG_ID")
+            or os.getenv("OPENAI_ORGANIZATION")
+        )
+        if values["openai_api_type"] in ("azure", "azure_ad", "azuread"):
+            raise ValueError(
```
Comment (efriis) [inline libs/partners/openai/langchain_openai/embeddings/base.py:211]: upgraded from a warning in new package

**Pull request merged**